File: C/palindrome.c
/**
 * Palindrome Checker - Multiple Methods
 * A palindrome is a word, phrase, number, or other sequence of characters
 * that reads the same forward and backward (ignoring spaces, punctuation, and capitalization).
 */

#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <stdlib.h>

/**
 * Helper function to remove spaces and convert to lowercase
 */
void cleanString(char* str, char* cleaned) {
    int j = 0;
    for (int i = 0; str[i] != '\0'; i++) {
        if (str[i] != ' ') {
            cleaned[j++] = tolower(str[i]);
        }
    }
    cleaned[j] = '\0';
}

/**
 * Method 1: Using string reversal
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */
int isPalindromeReverse(char* str) {
    char cleaned[1000];
    cleanString(str, cleaned);
    
    int len = strlen(cleaned);
    char reversed[1000];
    
    for (int i = 0; i < len; i++) {
        reversed[i] = cleaned[len - 1 - i];
    }
    reversed[len] = '\0';
    
    return strcmp(cleaned, reversed) == 0;
}

/**
 * Method 2: Using two pointers (efficient)
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
int isPalindromeTwoPointer(char* str) {
    char cleaned[1000];
    cleanString(str, cleaned);
    
    int left = 0;
    int right = strlen(cleaned) - 1;
    
    while (left < right) {
        if (cleaned[left] != cleaned[right]) {
            return 0;
        }
        left++;
        right--;
    }
    return 1;
}

/**
 * Method 3: Using recursion
 * Time Complexity: O(n)
 * Space Complexity: O(n) - due to recursion stack
 */
int checkPalindromeRecursive(char* str, int left, int right) {
    if (left >= right) {
        return 1;
    }
    if (str[left] != str[right]) {
        return 0;
    }
    return checkPalindromeRecursive(str, left + 1, right - 1);
}

int isPalindromeRecursive(char* str) {
    char cleaned[1000];
    cleanString(str, cleaned);
    return checkPalindromeRecursive(cleaned, 0, strlen(cleaned) - 1);
}

/**
 * Method 4: Using character array comparison
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
int isPalindromeCharArray(char* str) {
    char cleaned[1000];
    cleanString(str, cleaned);
    
    int len = strlen(cleaned);
    
    for (int i = 0; i < len / 2; i++) {
        if (cleaned[i] != cleaned[len - 1 - i]) {
            return 0;
        }
    }
    return 1;
}

/**
 * Method 5: Using dynamic memory allocation
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */
int isPalindromeDynamic(char* str) {
    // Clean the string
    int len = strlen(str);
    char* cleaned = (char*)malloc((len + 1) * sizeof(char));
    int j = 0;
    
    for (int i = 0; i < len; i++) {
        if (str[i] != ' ') {
            cleaned[j++] = tolower(str[i]);
        }
    }
    cleaned[j] = '\0';
    
    int cleanedLen = strlen(cleaned);
    int isPalindrome = 1;
    
    for (int i = 0; i < cleanedLen / 2; i++) {
        if (cleaned[i] != cleaned[cleanedLen - 1 - i]) {
            isPalindrome = 0;
            break;
        }
    }
    
    free(cleaned);
    return isPalindrome;
}

// Test examples
int main() {
    char* testCases[] = {
        "racecar",
        "hello",
        "A man a plan a canal Panama",
        "madam",
        "clang",
        "12321",
        "12345"
    };
    
    int numTests = sizeof(testCases) / sizeof(testCases[0]);
    
    printf("Palindrome Checker - Testing All Methods\n\n");
    printf("------------------------------------------------------------\n");
    
    for (int i = 0; i < numTests; i++) {
        printf("\nTesting: '%s'\n", testCases[i]);
        printf("  Method 1 (Reverse):       %s\n", isPalindromeReverse(testCases[i]) ? "true" : "false");
        printf("  Method 2 (Two Pointer):   %s\n", isPalindromeTwoPointer(testCases[i]) ? "true" : "false");
        printf("  Method 3 (Recursive):     %s\n", isPalindromeRecursive(testCases[i]) ? "true" : "false");
        printf("  Method 4 (Char Array):    %s\n", isPalindromeCharArray(testCases[i]) ? "true" : "false");
        printf("  Method 5 (Dynamic):       %s\n", isPalindromeDynamic(testCases[i]) ? "true" : "false");
    }
    
    return 0;
}

File: Python/palindrome.py
"""
Palindrome Checker - Multiple Methods
A palindrome is a word, phrase, number, or other sequence of characters 
that reads the same forward and backward (ignoring spaces, punctuation, and capitalization).
"""


def is_palindrome_reverse(s):
    """
    Method 1: Using string reversal
    Time Complexity: O(n)
    Space Complexity: O(n)
    """
    # Remove spaces and convert to lowercase
    s = s.replace(" ", "").lower()
    return s == s[::-1]


def is_palindrome_two_pointer(s):
    """
    Method 2: Using two pointers (efficient)
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    # Remove spaces and convert to lowercase
    s = s.replace(" ", "").lower()
    left = 0
    right = len(s) - 1
    
    while left < right:
        if s[left] != s[right]:
            return False
        left += 1
        right -= 1
    
    return True


def is_palindrome_recursive(s):
    """
    Method 3: Using recursion
    Time Complexity: O(n)
    Space Complexity: O(n) - due to recursion stack
    """
    # Remove spaces and convert to lowercase
    s = s.replace(" ", "").lower()
    
    # Helper recursive function
    def check_palindrome(s, left, right):
        if left >= right:
            return True
        if s[left] != s[right]:
            return False
        return check_palindrome(s, left + 1, right - 1)
    
    return check_palindrome(s, 0, len(s) - 1)


def is_palindrome_builtin(s):
    """
    Method 4: Using built-in reversed() function
    Time Complexity: O(n)
    Space Complexity: O(n)
    """
    # Remove spaces and convert to lowercase
    s = s.replace(" ", "").lower()
    return s == ''.join(reversed(s))


def is_palindrome_list_compare(s):
    """
    Method 5: Using list comparison
    Time Complexity: O(n)
    Space Complexity: O(n)
    """
    # Remove spaces and convert to lowercase
    s = s.replace(" ", "").lower()
    return list(s) == list(reversed(s))


# Test examples
if __name__ == "__main__":
    test_cases = [
        "racecar",
        "hello",
        "A man a plan a canal Panama",
        "madam",
        "python",
        "12321",
        "12345"
    ]
    
    print("Palindrome Checker - Testing All Methods\n")
    print("-" * 60)
    
    for test in test_cases:
        print(f"\nTesting: '{test}'")
        print(f"  Method 1 (Reverse):      {is_palindrome_reverse(test)}")
        print(f"  Method 2 (Two Pointer):  {is_palindrome_two_pointer(test)}")
        print(f"  Method 3 (Recursive):    {is_palindrome_recursive(test)}")
        print(f"  Method 4 (Built-in):     {is_palindrome_builtin(test)}")
        print(f"  Method 5 (List Compare): {is_palindrome_list_compare(test)}")
